packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -299,6 +254,55 @@ function handleComplexSelector(element: Element, selector: Complex): Element[] {
   return elements;
 }
 
+/**
+ * Handles compound selectors or a list of selectors for subjective element.
+ * To handle multi-paths matching, this function create branches internally and validates each cases.
+ * @param element The subjective element
+ * @param selectors A list of selectors to sequentially validated
+ * @returns A list of elements matches the list of selectors
+ */
+function handleCompoundSelector(element: Element, selectors: AST[]): Element[] {
+  if (selectors.length === 0) {
+    return [];
+  }
+
+  // Start with validating the subjective element with given selectors
+  const branches: [Element, number][] = querySelectorAll(element, selectors[0]).map(
+    function (element) {
+      return [element, 1];
+    },
+  );
+  const results: Element[] = [];
+
+  while (branches.length) {
+    const branch = branches.pop()!;
+    const element = branch[0];
+    let currentIndex = branch[1];
+    for (let candidates: Element[]; currentIndex < selectors.length; currentIndex++) {
+      // Handle transpose
+      if ((candidates = transpose(element, selectors[currentIndex])).length !== 0) {
```

Comment:
One aspect of transpose is that there are two situation where it results in an empty array:
* If it was a navigation (upwind), but did not find anything
* If it was a non-navigation (and exited immediately)

Both paths will then be unified here, and continue with the handle consecutive selectors matching. I wonder if it does not create problems for the case where upwind does not find anything.

Reply:
I was also got confused that part while implementing. I should have put a comment regarding that part but I forgot to do it as I aimed to make it atomic as possible.

Your understanding is correct towards the concern.

Reply:
I'm proposing to change the API of "transpose" to distinguish both cases. And then if the can exit early if there was a matching operator (:upward), but no match.

Should make it clearer, but also eliminate any nasty surprises later when the extra "matches" call would introduce.